packages/xstate-react/src/useMachine.ts

```diff
@@ -163,44 +166,47 @@ export function useMachine<
   >([]);
 
   useLayoutEffect(() => {
-    service.onTransition((currentState) => {
-      if (currentState.changed) {
-        setState(currentState);
-      }
-
-      if (currentState.actions.length) {
-        const reactEffectActions = currentState.actions.filter(
-          (action): action is ReactActionObject<TContext, TEvent> => {
-            return (
-              typeof action.exec === 'function' &&
-              '__effect' in (action as ReactActionObject<TContext, TEvent>).exec
-            );
-          }
-        );
-
-        const [effectActions, layoutEffectActions] = partition(
-          reactEffectActions,
-          (action): action is ReactActionObject<TContext, TEvent> => {
-            return action.exec.__effect === ReactEffectType.Effect;
-          }
-        );
-
-        effectActionsRef.current.push(
-          ...effectActions.map<ActionStateTuple<TContext, TEvent>>(
-            (effectAction) => [effectAction, currentState]
-          )
-        );
-
-        layoutEffectActionsRef.current.push(
-          ...layoutEffectActions.map<ActionStateTuple<TContext, TEvent>>(
-            (layoutEffectAction) => [layoutEffectAction, currentState]
-          )
-        );
-      }
-    });
-
-    // if service.state has not changed React should just bail out from this update
-    setState(service.state);
+    service
+      .onTransition((currentState) => {
+        // Only change the current state if:
+        // - the incoming state is the initial state (since it might have new actors)
+        // - OR the incoming state actually changed
+        if (currentState.changed || currentState.changed === undefined) {
```

Comment:
q: `|| currentState.changed === undefined` when this matters? if this matters in some situation then this should be covered by a test (and I don't think any of those that were added here cover this), seems like an easy thing to break by accident in the future

Reply:
It does matter because it would represent the second "live" initial state, and it's covered by the immediately-available test.

Reply:
oh, ok - it wasnt that clear to me as per: https://github.com/davidkpiano/xstate/pull/1239#discussion_r445579359 , I think that split mentioned there would make this clearer